File: C#/22.cs
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        IList<string> result = new List<string>();
        Generate(n, "", result, 0, 0);
        return result;
    }

    private void Generate(int n, string current, IList<string> result, int open, int close) {
        if (current.Length == n * 2) {
            result.Add(current);
            return;
        }

        if (open < n) {
            Generate(n, current + "(", result, open + 1, close);
        }
        if (close < open) {
            Generate(n, current + ")", result, open, close + 1);
        }
    }
}

File: python/22.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        
        def backtrack(s, left, right):
            if len(s) == 2*n:
                result.append(s)
                return
            if left < n:
                backtrack(s+"(", left+1, right)
            if right < left:
                backtrack(s+")", left, right+1)
        
        backtrack("", 0, 0)
        return result